content/concepts/multiplex/early-negotiation.md

```diff
@@ -0,0 +1,49 @@
+---
+title: "Early Multiplexer Negotiation"
+description: "Peers can use security protocol extenstions for early muxer negotiation."
+weight: 162
+---
+
+## Typical connection upgrade process
+
+Peers upgrade raw transport connections by using the same
+[multistream-selection](https://github.com/multiformats/multistream-select)
+protocol to negotiate security and stream multiplexing.
+
+First, security is established, and a security handshake is performed
+either for [Noise](../secure-comm/noise) or [TLS 1.3](../secure-comm/tls).
+Multistream-select will run again on top of
+the secure channel to negotiate a steam muxer, like [yamux](yamux).
+
+<!-- ADD DIAGRAM -->
+
+## Early muxer negotiation
+
+Early muxer negotiation is possible through the handshake of security protocols being able to negotiate higher-level
+protocols. The early negotiation takes place as a list of supported stream muxers is shared during the security protocol
+handshake, and a security protocol extension handles the stream muxer negotiation while it negotiates the secure channel
+establishment. This saves 1 RTT during the libp2p handshake and, as a result, reduces the TTFB (time to first byte).
```

Comment:
```suggestion
Early muxer negotiation is possible in the handshake of the TLS and Noise security protocols because each supports sending higher-level metadata.
For libp2p's purposes this metadata is composed as a list of stream muxers protocol IDs.
This list is shared in a security protocol "extension" during the handshake. 
For a libp2p node that supports both `mplex` and `yamux`, the list will look like this:
`[ "/yamux/1.0.0", "/mplex/6.7.0", "libp2p" ]`

Sharing this list in as a part of the cryptographic handshake payload allows for stream multiplexer selection to happen simultaneously.
This saves 1 RTT during the libp2p handshake and, as a result, reduces the TTFB (time to first byte).

Currently this feature is only supported in go-libp2p and for transports that do not have native encryption or multiplexing (TCP and WebSocket).

Read below for details on how this works in TLS and Noise.
```

Reply:
I like the suggestion, but we probably don't need to go into detail how the list looks like. It's different for TLS and for Noise (in Noise, we don't need the "libp2p" entry).

Reply:
I modified it to 
>For example, if a libp2p node supports both `mplex` and `yamux`, the list it advertises in would contain both of those protocols.

Reply:
```suggestion
Early muxer negotiation is a feature in libp2p that allows for the simultaneous selection of a stream multiplexer during the cryptographic handshake of the TLS and Noise security protocols. This is achieved by sharing a list of supported muxer protocols as a part of the handshake payload. 

For example, if a libp2p node supports mplex and yamux, it will advertise both in the list. This eliminates an extra round trip, improving TTFB (time to first byte) in the libp2p handshake. Currently, this feature is only supported in go-libp2p and for TCP and WebSocket transports that do not have native encryption or multiplexing.
```

Reply:
lgtm